# **🏢 Amazon ML Challenge 2026 - Business Entity Resolution**

This project focuses on matching business entity records across noisy, independent data sources with no shared unique IDs.

In this notebook:

# **🧹 01. Data Audit, Cleaning & Normalization Pipeline**

This notebook handles data ingestion, structural verification, noise distribution analysis, and text standardization across Source 1 (reference), Source 2, and Source 3 before candidate generation (blocking).

## **1. Problem Definition**

Business identity data arrives from disparate sources with typos, legal suffixes, missing street details, and varying regional address formats.

Given raw business records across 3 TSV files, normalize text components so that downstream blocking rules can group candidate pairs accurately without dropping true matches.



## **2. Data Architecture & Column Mapping**

All datasets are tab-separated (`.tsv`) to safely handle commas inside address strings.

**Key Datasets:**
- `train_source1.tsv`: Reference deduplicated records (`S1-` prefix)
- `train_source2.tsv`: Query/fragment source 1 (`S2-` prefix)
- `train_source3.tsv`: Query/fragment source 2 (`S3-` prefix)
- `train_ground_truth.tsv`: Ground truth mapping (`source1_entity_id` → matched `S2`/`S3` IDs)

**Schema Across All Source Files:**

| **Column** | **Type** | **Description** | **Common Noise / Challenges** |
|---|---|---|---|
| `entity_id` | String | Unique record ID (`S1-xxx`, `S2-xxx`, `S3-xxx`) | Source prefix indicator |
| `business_name` | String | Name of the business | Legal suffixes (LLC, Pvt Ltd, Inc), typos, word reordering |
| `business_address` | String | Street address / location | Missing values, landmark references, abbreviations (`Rd`, `St`) |
| `country` | String | Country code (`US`, `India` in Train; `France` in Test) | Domain shift across regions |



## **3. Pipeline Execution Steps**

1. TSV Integrity & Ingestion: Verify tab separation and memory usage.
2. Missingness & Country Coverage Audit: Check null percentages for names and addresses across US vs India.
3. String Normalization Engine: Lowercasing, punctuation stripping, whitespace collapsing, and legal suffix stripping.
4. Digit & Token Extraction: Isolate street numbers/zip codes and sorted name tokens for blocking keys.

```python
Raw TSV file
     |
     v
Step 1: Does it load correctly? (integrity check)
     |
     v
Step 2: What's missing, and does it differ by country?
     |
     v
Step 3: Clean text (same business should look the same)
     |
     v
Step 4: Pull out blocking ingredients (zip, number, name tokens)
     |
     v
Ready to actually build bucket logic
```

### **1. Loading and testing the dataset**

In [ ]:
import pandas as pd
import numpy as np
from pathlib import Path

# Set up data directory path
DATA_DIR = Path("../data/student_resource/dataset/train")

# ---------------------------------------------------------
# STEP 1: TSV Integrity & Ingestion
# ---------------------------------------------------------
print("Loading datasets...")

df_s1 = pd.read_csv(DATA_DIR / "train_source1.tsv", sep="\t")
df_s2 = pd.read_csv(DATA_DIR / "train_source2.tsv", sep="\t")
df_s3 = pd.read_csv(DATA_DIR / "train_source3.tsv", sep="\t")
df_gt = pd.read_csv(DATA_DIR / "train_ground_truth.tsv", sep="\t")

print("Ingestion complete!\n")
print("Ingestion2 complete!\n")

Loading datasets...
Ingestion complete!



In [ ]:
df_s1.head()

,entity_id,business_name,business_address,country
0,S1-925783039,Orelee's Barbershop,"1795 Westchester Drive, High Point, NC",US
1,S1-773889195,Prime Money,"17560 Ellis Road, Tahlequah, OK",US
2,S1-377745466,B+ Retail Inc,"1712 Montebello Avenue, Phoenix, AZ",US
3,S1-133037285,Christ Chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",US
4,S1-755362802,Prabhav Business Center,"797, Lake Town Block A, Kolkata, Howrah, West ...",India


In [ ]:
df_s2.head()

,entity_id,business_name,business_address,country
0,S2-166376419,राम मार्केटिंग प्राइवेट लिमिटेड,"KH NO. -570/13, NEW DELHI, WEST DELHI, Delhi",India
1,S2-764573417,-- Holloway Peak Inc Seafood,"105 ELM ST, MORGANTON, NC",US
2,S2-639257739,आदित्य प्रॉपर्टीज एलएलपी,"G-3/571, GULMOHAR COLONY, BHOPAL, Madhya Pradesh",India
3,S2-163963287,Summit Inc,"GREENSBORO, NC, 19 1/2 STARDUST TRAIL",US
4,S2-49942811,Delta Tetlecommunication Inc,"914 PIERPONT AVE, CLEVELAND, OH",US


In [ ]:
df_s3.head()

,entity_id,business_name,business_address,country
0,S3-202863386,wilfordhancock.com,"Mack Rd, Haltom City, Texas",US
1,S3-859268022,International South Consultants Private Ltd,NaN,India
2,S3-22467283,LLC Moncada Léarning Center,"5780 Fawn Ct, Fort Worth, Texas",US
3,S3-671162755,Moyna's Coffee,"1 Ivanhoe Ave, PO Box 6009, Cincinnati, Ohio",US
4,S3-960981775,Pvt. EFS Print Ventures Ltd.,"Door No 183, 41St Cross, 22Nd Main 9Th Block J...",India


In [ ]:
df_gt.head()

,source1_entity_id,matched_entity_ids
0,S1-965667,"S2-681193310,S2-743505751,S3-775321672,S3-1129..."
1,S1-55344266,"S2-249013014,S2-197070651,S3-478195123,S3-3843..."
2,S1-343815751,"S2-790675320,S2-479876582,S3-878454467"
3,S1-656753428,"S2-153058913,S2-24659151,S3-679606215"
4,S1-102811957,"S2-478959098,S2-553508714,S2-625774905,S3-7280..."


In [ ]:
df_s1.isnull().sum()

entity_id           0
business_name       0
business_address    0
country             0
dtype: int64

In [ ]:
df_s2.isnull().sum()

entity_id                0
business_name            2
business_address    168967
country                  0
dtype: int64

In [ ]:
pct = (df_s2.isnull().sum() / len(df_s2)) * 100
pct

entity_id           0.000000
business_name       0.000040
business_address    3.356105
country             0.000000
dtype: float64

In [ ]:
df_s3.isnull().sum()

entity_id                0
business_name           13
business_address    175916
country                  0
dtype: int64

In [ ]:
df_gt.isnull().sum()

source1_entity_id          0
matched_entity_ids    123247
dtype: int64

In [ ]:
print(len(df_s1))
print(len(df_s2))
print(len(df_s3))
print(len(df_gt))

2206821
5034616
5285603
2206821


In [ ]:
# Function to calculate missingness summary
def audit_missingness(df, name):
    missing = df.isnull().sum()
    pct = (df.isnull().sum() / len(df)) * 100
    audit_df = pd.DataFrame({'Missing Count': missing, 'Missing %': pct.round(2)})
    print(f"--- Missingness Audit: {name} ---")
    print(audit_df)
    print("\nCountry Breakdown:")
    print(df['country'].value_counts(dropna=False))
    print("=" * 45 + "\n")

audit_missingness(df_s1, "Source 1 (S1)")
audit_missingness(df_s2, "Source 2 (S2)")
audit_missingness(df_s3, "Source 3 (S3)")

--- Missingness Audit: Source 1 (S1) ---
                  Missing Count  Missing %
entity_id                     0        0.0
business_name                 0        0.0
business_address              0        0.0
country                       0        0.0

Country Breakdown:
country
US       1323633
India     883188
Name: count, dtype: int64

--- Missingness Audit: Source 2 (S2) ---
                  Missing Count  Missing %
entity_id                     0       0.00
business_name                 2       0.00
business_address         168967       3.36
country                       0       0.00

Country Breakdown:
country
US       3016817
India    2017799
Name: count, dtype: int64

--- Missingness Audit: Source 3 (S3) ---
                  Missing Count  Missing %
entity_id                     0       0.00
business_name                13       0.00
business_address         175916       3.33
country                       0       0.00

Country Breakdown:
country
US       3170056
India    

## **📊 Amazon ML Challenge 2026 — Data Audit & Normalization Summary**

### **1. Quick Dataset Overview**

| File | Total Records | Missing Names | Missing Addresses | Missing % (Addr) | US Count | India Count |
|---|---|---|---|---|---|---|
| `train_source1.tsv` | 2,206,821 | 0 | 0 | **0.0%** | 1,323,633 | 883,188 |
| `train_source2.tsv` | 5,034,616 | 2 | 168,967 | **3.36%** | 3,016,817 | 2,017,799 |
| `train_source3.tsv` | 5,285,603 | 13 | 175,916 | **3.33%** | 3,170,056 | 2,115,547 |

---

### **2. Key Audit Insights**

1. **Source 1 is Complete (Ground Reference):**
   - $S_1$ has zero missing values across all columns. It acts as our clean reference anchor.
   
2. **The Address Missingness Trap (~344,000 Records):**
   - Over **344,000 records** across $S_2$ and $S_3$ have no `business_address`.
   - **Pipeline Rule:** We **cannot** use address-only blocking keys. Any candidate generation rule requiring an address will permanently drop 3.3%+ of potential matches, lowering our maximum achievable recall.

3. **Geographic Distribution:**
   - Training set ratio is roughly **60% US / 40% India**.
   - Test set will introduce **France**. All cleaning and blocking logic must remain language-agnostic and handle French formats seamlessly.

---

## **3. Standardized Cleaning & Feature Strategy**

To ensure matching works across all sources, we generate three standardized representations per record:

1. `clean_name`: Lowercased, stripped punctuation, and stripped legal suffixes (US, India, and France: `inc`, `llc`, `pvt`, `ltd`, `sarl`, `sas`, `sa`, etc.).
2. `sorted_name_tokens`: Alphabetically ordered word tokens to handle word reordering (e.g., *"Metro Shoes Downtown"* $\to$ *"downtown metro shoes"*).
3. `address_digits`: Extracted numbers/zips from address strings to serve as hard numeric anchors independent of text typos.

---

In [ ]:
import re
import pandas as pd
import unicodedata
import time

# Combined Legal Suffixes: English (US/India), French (Test set), and Hindi Transliterations
LEGAL_SUFFIXES = r'\b(pvt|ltd|inc|llc|corp|corporation|co|limited|private|gmbh|sa|sarl|sas|sasu|eurl|sci|snc|gie|प्राइवेट|लिमिटेड|एलएलपी)\b'

def strip_latin_accents(text: str) -> str:
    """Converts French/Latin accents to base characters (e.g., 'é' -> 'e', 'ç' -> 'c')
    while keeping non-Latin scripts (like Hindi Devanagari) unaffected.
    """
    # NFKD decomposition separates base letters from combining accent marks
    nfkd_form = unicodedata.normalize('NFKD', text)
    # Remove only the combining accent marks (\p{Mn})
    return "".join([c for c in nfkd_form if not unicodedata.combining(c)])

def normalize_series(series: pd.Series) -> pd.Series:
    """Vectorized normalization that folds French accents and preserves Hindi script."""
    # 1. Fill NaNs
    s = series.fillna("")

    # 2. Lowercase
    s = s.str.lower()

    # 3. Strip French accents (e.g., 'société' -> 'societe') so accented & unaccented records match
    s = s.apply(strip_latin_accents)

    # 4. Strip punctuation BUT KEEP words (\w), spaces (\s), and Devanagari script (\u0900-\u097F)
    s = s.str.replace(r'[^\w\s\u0900-\u097F]', ' ', regex=True)

    # 5. Strip legal corporate suffixes (English, French, Hindi)
    s = s.str.replace(LEGAL_SUFFIXES, ' ', regex=True)

    # 6. Collapse multiple spaces and trim
    s = s.str.replace(r'\s+', ' ', regex=True).str.strip()

    return s

# Time the execution to ensure it doesn't hang
print("Normalizing Source 1...")
start = time.time()
df_s1['clean_name'] = normalize_series(df_s1['business_name'])
df_s1['clean_address'] = normalize_series(df_s1['business_address'])
print(f"Source 1 done in {time.time() - start:.2f} seconds\n")

print("Normalizing Source 2 (The 5M row beast)...")
start = time.time()
df_s2['clean_name'] = normalize_series(df_s2['business_name'])
df_s2['clean_address'] = normalize_series(df_s2['business_address'])
print(f"Source 2 done in {time.time() - start:.2f} seconds\n")

print("Normalizing Source 3...")
start = time.time()
df_s3['clean_name'] = normalize_series(df_s3['business_name'])
df_s3['clean_address'] = normalize_series(df_s3['business_address'])
print(f"Source 3 done in {time.time() - start:.2f} seconds")

Normalizing Source 1...
Source 1 done in 28.82 seconds

Normalizing Source 2 (The 5M row beast)...
Source 2 done in 68.24 seconds

Normalizing Source 3...
Source 3 done in 76.11 seconds


In [ ]:
df_s1.head()

,entity_id,business_name,business_address,country,clean_name,clean_address
0,S1-925783039,Orelee's Barbershop,"1795 Westchester Drive, High Point, NC",US,orelee s barbershop,1795 westchester drive high point nc
1,S1-773889195,Prime Money,"17560 Ellis Road, Tahlequah, OK",US,prime money,17560 ellis road tahlequah ok
2,S1-377745466,B+ Retail Inc,"1712 Montebello Avenue, Phoenix, AZ",US,b retail,1712 montebello avenue phoenix az
3,S1-133037285,Christ Chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",US,christ chapel,2100 cameron drive unit apartment g dundalk md
4,S1-755362802,Prabhav Business Center,"797, Lake Town Block A, Kolkata, Howrah, West ...",India,prabhav business center,797 lake town block a kolkata howrah west bengal


In [ ]:
df_s2.head()

,entity_id,business_name,business_address,country,clean_name,clean_address
0,S2-166376419,राम मार्केटिंग प्राइवेट लिमिटेड,"KH NO. -570/13, NEW DELHI, WEST DELHI, Delhi",India,राम मारकेटिंग पराइवेट,kh no 570 13 new delhi west delhi delhi
1,S2-764573417,-- Holloway Peak Inc Seafood,"105 ELM ST, MORGANTON, NC",US,holloway peak seafood,105 elm st morganton nc
2,S2-639257739,आदित्य प्रॉपर्टीज एलएलपी,"G-3/571, GULMOHAR COLONY, BHOPAL, Madhya Pradesh",India,आदितय परॉपरटीज एलएलपी,g 3 571 gulmohar colony bhopal madhya pradesh
3,S2-163963287,Summit Inc,"GREENSBORO, NC, 19 1/2 STARDUST TRAIL",US,summit,greensboro nc 19 1 2 stardust trail
4,S2-49942811,Delta Tetlecommunication Inc,"914 PIERPONT AVE, CLEVELAND, OH",US,delta tetlecommunication,914 pierpont ave cleveland oh


In [ ]:
df_s3.head()

,entity_id,business_name,business_address,country,clean_name,clean_address
0,S3-202863386,wilfordhancock.com,"Mack Rd, Haltom City, Texas",US,wilfordhancock com,mack rd haltom city texas
1,S3-859268022,International South Consultants Private Ltd,NaN,India,international south consultants,
2,S3-22467283,LLC Moncada Léarning Center,"5780 Fawn Ct, Fort Worth, Texas",US,moncada learning center,5780 fawn ct fort worth texas
3,S3-671162755,Moyna's Coffee,"1 Ivanhoe Ave, PO Box 6009, Cincinnati, Ohio",US,moyna s coffee,1 ivanhoe ave po box 6009 cincinnati ohio
4,S3-960981775,Pvt. EFS Print Ventures Ltd.,"Door No 183, 41St Cross, 22Nd Main 9Th Block J...",India,efs print ventures,door no 183 41st cross 22nd main 9th block jay...


In [ ]:
df_gt.head()

,source1_entity_id,matched_entity_ids
0,S1-965667,"S2-681193310,S2-743505751,S3-775321672,S3-1129..."
1,S1-55344266,"S2-249013014,S2-197070651,S3-478195123,S3-3843..."
2,S1-343815751,"S2-790675320,S2-479876582,S3-878454467"
3,S1-656753428,"S2-153058913,S2-24659151,S3-679606215"
4,S1-102811957,"S2-478959098,S2-553508714,S2-625774905,S3-7280..."


In [ ]:
import time

print("Saving cleaned datasets to Parquet format using fastparquet...")
start = time.time()

# We added engine='fastparquet' to bypass the PyArrow bug
df_s1.to_parquet("../data/clean_source1.parquet", index=False, engine='fastparquet')
print("Source 1 saved.")

df_s2.to_parquet("../data/clean_source2.parquet", index=False, engine='fastparquet')
print("Source 2 saved.")

df_s3.to_parquet("../data/clean_source3.parquet", index=False, engine='fastparquet')
print(f"Source 3 saved. All files exported in {time.time() - start:.2f} seconds.")

Saving cleaned datasets to Parquet format using fastparquet...
Source 1 saved.
Source 2 saved.
Source 3 saved. All files exported in 51.27 seconds.


## **🧹 Data Cleaning & Normalization Handoff**

The initial data ingestion and text normalization phase is complete. The datasets are cleaned, standardized, and ready for the candidate generation (blocking) phase.

### **1. What Was Added**
- **New Columns:** Added `clean_name` and `clean_address` to all three datasets ($S_1$, $S_2$, $S_3$).
- **Preserved Raw Data:** The original `business_name` and `business_address` columns remain untouched for reference or fallback.

### **2. What Was Removed & Why**
- **French/Latin Accents Folded:** Accents were converted to base letters (e.g., `Société` $\to$ `societe`) using Unicode decomposition. *Why:* Prevents matching failures on the France test set when sources inconsistently use accents.
- **Punctuation Stripped (Safely):** Removed commas, periods, and symbols, but strictly preserved English/French alphanumeric characters (`\w`) AND the entire Hindi Devanagari unicode block (`\u0900-\u097F`). *Why:* Standard regex destroys Hindi matras (vowels). This custom regex ensures Hindi text stays 100% intact.
- **Legal Corporate Suffixes Removed:** Stripped high-frequency legal fluff across all three regions (e.g., `inc`, `llc`, `pvt`, `ltd`, `sarl`, `sas`, `प्राइवेट`, `लिमिटेड`). *Why:* These suffixes are noise. Stripping them ensures the core brand name is isolated for exact-match blocking and similarity scoring.
- **Whitespace Collapsed:** Multiple spaces and trailing whitespaces were trimmed into single spaces.

### **3. How It Is Saved**
The cleaned DataFrames were exported to the `data/clean_data/` folder as **Parquet files**, not TSVs.
- `clean_source1.parquet`
- `clean_source2.parquet`
- `clean_source3.parquet`

*Why Parquet:* It reads/writes significantly faster than TSV, uses less disk space, and critically, it preserves empty strings `""` instead of accidentally converting them back into `NaN` values like pandas sometimes does with CSV/TSV.

### **4. How to Load It (For the Blocking Team)**
You can load these files instantly into your blocking script using `pandas`.

```python
import pandas as pd

# Load the pre-cleaned datasets
df_s1 = pd.read_parquet("../data/clean_data/clean_source1.parquet")
df_s2 = pd.read_parquet("../data/clean_data/clean_source2.parquet")
df_s3 = pd.read_parquet("../data/clean_data/clean_source3.parquet")

# Now you can use clean_name and clean_address to build your blocking keys
# (e.g., generating sorted tokens or extracting address digits)

In [4]:
%cd /content/Amazon-ML-Challenge

/content/Amazon-ML-Challenge


In [5]:
!pwd

/content/Amazon-ML-Challenge


In [6]:
import os

print("Current working directory:")
print(os.getcwd())

print("\nRepository contents:")
print(os.listdir("/content/Amazon-ML-Challenge"))

print("\nSearching for dataset folders:")
!find /content -maxdepth 5 -type d -name "dataset" 2>/dev/null

Current working directory:
/content/Amazon-ML-Challenge

Repository contents:
['.gitignore', '.git', 'code', 'utils', 'notebooks', 'README.md', 'DATASET_INVENTORY.md', 'LICENSE', 'Documentation_template.md', 'submissions.md']

Searching for dataset folders:


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [14]:
!find "/content/drive/MyDrive" -type f -name "clean_data.zip" 2>/dev/null

/content/drive/MyDrive/clean_data.zip


In [15]:
import zipfile

ZIP_PATH = "/content/drive/MyDrive/clean_data.zip"

with zipfile.ZipFile(ZIP_PATH, "r") as z:
    print("\n".join(z.namelist()))

clean_data/
clean_data/clean_source1.parquet
clean_data/clean_source2.parquet
clean_data/clean_source3.parquet


In [16]:
from pathlib import Path
import zipfile

ZIP_PATH = Path("/content/drive/MyDrive/clean_data.zip")
PROJECT_DIR = Path("/content/Amazon-ML-Challenge")
DATA_DIR = PROJECT_DIR / "data"

DATA_DIR.mkdir(parents=True, exist_ok=True)

with zipfile.ZipFile(ZIP_PATH, "r") as z:
    z.extractall(DATA_DIR)

print("Extraction complete.")
print("Files:")
for f in (DATA_DIR / "clean_data").glob("*"):
    print(f)

Extraction complete.
Files:
/content/Amazon-ML-Challenge/data/clean_data/clean_source1.parquet
/content/Amazon-ML-Challenge/data/clean_data/clean_source2.parquet
/content/Amazon-ML-Challenge/data/clean_data/clean_source3.parquet


In [18]:
from pathlib import Path
import pyarrow.parquet as pq

DATA_DIR = Path("/content/Amazon-ML-Challenge/data/clean_data")

for name in ["clean_source1.parquet", "clean_source2.parquet", "clean_source3.parquet"]:
    path = DATA_DIR / name
    pf = pq.ParquetFile(path)

    print(f"\n{name}")
    print("-" * 60)
    print("Rows:", pf.metadata.num_rows)
    print("Columns:", pf.schema_arrow.names)
    print("Row groups:", pf.num_row_groups)


clean_source1.parquet
------------------------------------------------------------
Rows: 2206821
Columns: ['entity_id', 'business_name', 'business_address', 'country', 'clean_name', 'clean_address']
Row groups: 1

clean_source2.parquet
------------------------------------------------------------
Rows: 5034616
Columns: ['entity_id', 'business_name', 'business_address', 'country', 'clean_name', 'clean_address']
Row groups: 1

clean_source3.parquet
------------------------------------------------------------
Rows: 5285603
Columns: ['entity_id', 'business_name', 'business_address', 'country', 'clean_name', 'clean_address']
Row groups: 1


In [20]:
import pandas as pd


In [22]:
!pip -q install duckdb

import duckdb

DATA_DIR = "/content/Amazon-ML-Challenge/data/clean_data"

con = duckdb.connect()

sources = [
    "clean_source1",
    "clean_source2",
    "clean_source3"
]

# ============================================================
# 1. BASIC DATA PROFILE
# ============================================================

print("\n" + "="*90)
print("BASIC DATA PROFILE")
print("="*90)

for source in sources:
    path = f"{DATA_DIR}/{source}.parquet"

    result = con.sql(f"""
        SELECT
            COUNT(*) AS total_rows,
            COUNT(DISTINCT entity_id) AS unique_entity_ids,
            COUNT(DISTINCT clean_name) AS unique_clean_names,
            COUNT(DISTINCT clean_address) AS unique_clean_addresses,
            COUNT(DISTINCT country) AS unique_countries,

            COUNT(*) FILTER (
                WHERE clean_name IS NULL OR TRIM(clean_name) = ''
            ) AS empty_clean_names,

            COUNT(*) FILTER (
                WHERE clean_address IS NULL OR TRIM(clean_address) = ''
            ) AS empty_clean_addresses
        FROM read_parquet('{path}')
    """).fetchone()

    print(f"\n{source}")
    print("-" * 50)
    print(f"Total rows           : {result[0]:,}")
    print(f"Unique entity IDs    : {result[1]:,}")
    print(f"Unique clean names   : {result[2]:,}")
    print(f"Unique clean address : {result[3]:,}")
    print(f"Unique countries     : {result[4]:,}")
    print(f"Empty clean names    : {result[5]:,}")
    print(f"Empty clean addresses: {result[6]:,}")


# ============================================================
# 2. COUNTRY DISTRIBUTION
# ============================================================

for source in sources:
    path = f"{DATA_DIR}/{source}.parquet"

    print("\n" + "="*90)
    print(f"{source.upper()} — COUNTRY DISTRIBUTION")
    print("="*90)

    print(
        con.sql(f"""
            SELECT
                country,
                COUNT(*) AS records
            FROM read_parquet('{path}')
            GROUP BY country
            ORDER BY records DESC
            LIMIT 30
        """)
    )


# ============================================================
# 3. HOW MANY RECORDS WOULD MATCH USING:
#    clean_name + country
# ============================================================

print("\n" + "="*90)
print("BLOCK A: clean_name + country")
print("="*90)

s1_path = f"{DATA_DIR}/clean_source1.parquet"
s2_path = f"{DATA_DIR}/clean_source2.parquet"
s3_path = f"{DATA_DIR}/clean_source3.parquet"

for target_name, target_path in [
    ("S2", s2_path),
    ("S3", s3_path)
]:

    result = con.sql(f"""
        WITH s1_keys AS (
            SELECT
                clean_name,
                country,
                COUNT(*) AS s1_count
            FROM read_parquet('{s1_path}')
            WHERE clean_name IS NOT NULL
              AND TRIM(clean_name) != ''
            GROUP BY clean_name, country
        ),
        target_keys AS (
            SELECT
                clean_name,
                country,
                COUNT(*) AS target_count
            FROM read_parquet('{target_path}')
            WHERE clean_name IS NOT NULL
              AND TRIM(clean_name) != ''
            GROUP BY clean_name, country
        )
        SELECT
            COUNT(*) AS matched_keys,
            SUM(s1_count) AS s1_records_in_matching_blocks,
            SUM(target_count) AS target_records_in_matching_blocks,
            SUM(s1_count * target_count) AS total_candidate_pairs
        FROM s1_keys s
        INNER JOIN target_keys t
            ON s.clean_name = t.clean_name
           AND s.country = t.country
    """).fetchone()

    print(f"\nS1 → {target_name}")
    print(f"Matching keys       : {result[0]:,}")
    print(f"S1 records affected : {result[1]:,}")
    print(f"{target_name} records   : {result[2]:,}")
    print(f"Candidate pairs     : {result[3]:,}")


# ============================================================
# 4. BLOCK B: clean_address + country
# ============================================================

print("\n" + "="*90)
print("BLOCK B: clean_address + country")
print("="*90)

for target_name, target_path in [
    ("S2", s2_path),
    ("S3", s3_path)
]:

    result = con.sql(f"""
        WITH s1_keys AS (
            SELECT
                clean_address,
                country,
                COUNT(*) AS s1_count
            FROM read_parquet('{s1_path}')
            WHERE clean_address IS NOT NULL
              AND TRIM(clean_address) != ''
            GROUP BY clean_address, country
        ),
        target_keys AS (
            SELECT
                clean_address,
                country,
                COUNT(*) AS target_count
            FROM read_parquet('{target_path}')
            WHERE clean_address IS NOT NULL
              AND TRIM(clean_address) != ''
            GROUP BY clean_address, country
        )
        SELECT
            COUNT(*) AS matched_keys,
            SUM(s1_count) AS s1_records_in_matching_blocks,
            SUM(target_count) AS target_records_in_matching_blocks,
            SUM(s1_count * target_count) AS total_candidate_pairs
        FROM s1_keys s
        INNER JOIN target_keys t
            ON s.clean_address = t.clean_address
           AND s.country = t.country
    """).fetchone()

    print(f"\nS1 → {target_name}")
    print(f"Matching keys       : {result[0]:,}")
    print(f"S1 records affected : {result[1]:,}")
    print(f"{target_name} records   : {result[2]:,}")
    print(f"Candidate pairs     : {result[3]:,}")


# ============================================================
# 5. NAME DUPLICATION — IMPORTANT FOR BLOCK A
# ============================================================

print("\n" + "="*90)
print("MOST COMMON clean_name + country COMBINATIONS")
print("="*90)

for source in sources:
    path = f"{DATA_DIR}/{source}.parquet"

    print(f"\n===== {source} =====")

    print(
        con.sql(f"""
            SELECT
                clean_name,
                country,
                COUNT(*) AS frequency
            FROM read_parquet('{path}')
            WHERE clean_name IS NOT NULL
              AND TRIM(clean_name) != ''
            GROUP BY clean_name, country
            ORDER BY frequency DESC
            LIMIT 20
        """)
    )


BASIC DATA PROFILE


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


clean_source1
--------------------------------------------------
Total rows           : 2,206,821
Unique entity IDs    : 2,206,821
Unique clean names   : 1,377,451
Unique clean address : 2,130,142
Unique countries     : 2
Empty clean names    : 8
Empty clean addresses: 0


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


clean_source2
--------------------------------------------------
Total rows           : 5,034,616
Unique entity IDs    : 5,034,616
Unique clean names   : 3,355,825
Unique clean address : 4,285,878
Unique countries     : 2
Empty clean names    : 35
Empty clean addresses: 168,967


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


clean_source3
--------------------------------------------------
Total rows           : 5,285,603
Unique entity IDs    : 5,285,603
Unique clean names   : 3,584,660
Unique clean address : 4,616,027
Unique countries     : 2
Empty clean names    : 108
Empty clean addresses: 175,916

CLEAN_SOURCE1 — COUNTRY DISTRIBUTION
┌─────────┬─────────┐
│ country │ records │
│ varchar │  int64  │
├─────────┼─────────┤
│ US      │ 1323633 │
│ India   │  883188 │
└─────────┴─────────┘


CLEAN_SOURCE2 — COUNTRY DISTRIBUTION
┌─────────┬─────────┐
│ country │ records │
│ varchar │  int64  │
├─────────┼─────────┤
│ US      │ 3016817 │
│ India   │ 2017799 │
└─────────┴─────────┘


CLEAN_SOURCE3 — COUNTRY DISTRIBUTION
┌─────────┬─────────┐
│ country │ records │
│ varchar │  int64  │
├─────────┼─────────┤
│ US      │ 3170056 │
│ India   │ 2115547 │
└─────────┴─────────┘


BLOCK A: clean_name + country


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


S1 → S2
Matching keys       : 902,035
S1 records affected : 1,705,606
S2 records   : 2,066,979
Candidate pairs     : 38,681,017


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


S1 → S3
Matching keys       : 913,142
S1 records affected : 1,719,136
S3 records   : 2,181,037
Candidate pairs     : 43,114,141

BLOCK B: clean_address + country


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


S1 → S2
Matching keys       : 354,153
S1 records affected : 385,075
S2 records   : 474,166
Candidate pairs     : 564,542


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


S1 → S3
Matching keys       : 170,995
S1 records affected : 197,084
S3 records   : 173,128
Candidate pairs     : 205,584

MOST COMMON clean_name + country COMBINATIONS

===== clean_source1 =====


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌────────────────────────┬─────────┬───────────┐
│       clean_name       │ country │ frequency │
│        varchar         │ varchar │   int64   │
├────────────────────────┼─────────┼───────────┤
│ meridian               │ US      │       522 │
│ cedar                  │ US      │       321 │
│ cascade                │ US      │       311 │
│ summit                 │ US      │       310 │
│ family center          │ US      │       305 │
│ helios                 │ US      │       299 │
│ ear nose throat group  │ US      │       295 │
│ granite                │ US      │       291 │
│ sapphire               │ US      │       291 │
│ sterling               │ US      │       290 │
│ lynx                   │ US      │       289 │
│ amber                  │ US      │       289 │
│ slate                  │ US      │       285 │
│ primary care group     │ US      │       283 │
│ citadel                │ US      │       283 │
│ anchor                 │ US      │       282 │
│ ember             

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌─────────────────────┬─────────┬───────────┐
│     clean_name      │ country │ frequency │
│       varchar       │ varchar │   int64   │
├─────────────────────┼─────────┼───────────┤
│ meridian            │ US      │       736 │
│ physical therapy    │ US      │       579 │
│ primary care        │ US      │       576 │
│ womens health       │ US      │       533 │
│ internal medicine   │ US      │       522 │
│ behavioral health   │ US      │       517 │
│ pediatric dental    │ US      │       513 │
│ urgent care         │ US      │       508 │
│ family center       │ US      │       505 │
│ pediatric dentistry │ US      │       496 │
│ ear nose throat     │ US      │       417 │
│ anchor              │ US      │       416 │
│ foot ankle          │ US      │       415 │
│ summit              │ US      │       413 │
│ cascade             │ US      │       410 │
│ helios              │ US      │       403 │
│ cedar               │ US      │       402 │
│ lynx                │ US      │ 

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌─────────────────────┬─────────┬───────────┐
│     clean_name      │ country │ frequency │
│       varchar       │ varchar │   int64   │
├─────────────────────┼─────────┼───────────┤
│ meridian            │ US      │       666 │
│ primary care        │ US      │       560 │
│ physical therapy    │ US      │       558 │
│ pediatric dental    │ US      │       545 │
│ family center       │ US      │       535 │
│ urgent care         │ US      │       526 │
│ pediatric dentistry │ US      │       508 │
│ womens health       │ US      │       506 │
│ behavioral health   │ US      │       475 │
│ internal medicine   │ US      │       469 │
│ summit              │ US      │       467 │
│ cascade             │ US      │       458 │
│ sterling            │ US      │       441 │
│ cedar               │ US      │       440 │
│ apex                │ US      │       436 │
│ cornerstone         │ US      │       419 │
│ vanguard            │ US      │       418 │
│ helios              │ US      │ 

In [4]:
import duckdb

DATA_DIR = "/content/Amazon-ML-Challenge/data/clean_data"

s1_path = f"{DATA_DIR}/clean_source1.parquet"
s2_path = f"{DATA_DIR}/clean_source2.parquet"
s3_path = f"{DATA_DIR}/clean_source3.parquet"

con = duckdb.connect()

query = f"""
WITH
s1 AS (
    SELECT entity_id, clean_name, clean_address, country
    FROM read_parquet('{s1_path}')
    WHERE clean_name IS NOT NULL
      AND TRIM(clean_name) != ''
      AND clean_address IS NOT NULL
      AND TRIM(clean_address) != ''
),

s2 AS (
    SELECT entity_id, clean_name, clean_address, country
    FROM read_parquet('{s2_path}')
    WHERE clean_name IS NOT NULL
      AND TRIM(clean_name) != ''
      AND clean_address IS NOT NULL
      AND TRIM(clean_address) != ''
),

s3 AS (
    SELECT entity_id, clean_name, clean_address, country
    FROM read_parquet('{s3_path}')
    WHERE clean_name IS NOT NULL
      AND TRIM(clean_name) != ''
      AND clean_address IS NOT NULL
      AND TRIM(clean_address) != ''
),

s1_keys AS (
    SELECT clean_name, clean_address, country, COUNT(*) AS n
    FROM s1
    GROUP BY clean_name, clean_address, country
),

s2_keys AS (
    SELECT clean_name, clean_address, country, COUNT(*) AS n
    FROM s2
    GROUP BY clean_name, clean_address, country
),

s3_keys AS (
    SELECT clean_name, clean_address, country, COUNT(*) AS n
    FROM s3
    GROUP BY clean_name, clean_address, country
),

positive_s2 AS (
    SELECT
        a.entity_id AS source1_entity_id,
        b.entity_id AS candidate_entity_id,
        'S2' AS source
    FROM s1 a
    JOIN s1_keys ak
      ON a.clean_name = ak.clean_name
     AND a.clean_address = ak.clean_address
     AND a.country = ak.country

    JOIN s2 b
      ON a.clean_name = b.clean_name
     AND a.clean_address = b.clean_address
     AND a.country = b.country

    JOIN s2_keys bk
      ON b.clean_name = bk.clean_name
     AND b.clean_address = bk.clean_address
     AND b.country = bk.country

    WHERE ak.n = 1
      AND bk.n = 1
),

positive_s3 AS (
    SELECT
        a.entity_id AS source1_entity_id,
        b.entity_id AS candidate_entity_id,
        'S3' AS source
    FROM s1 a
    JOIN s1_keys ak
      ON a.clean_name = ak.clean_name
     AND a.clean_address = ak.clean_address
     AND a.country = ak.country

    JOIN s3 b
      ON a.clean_name = b.clean_name
     AND a.clean_address = b.clean_address
     AND a.country = b.country

    JOIN s3_keys bk
      ON b.clean_name = bk.clean_name
     AND b.clean_address = bk.clean_address
     AND b.country = bk.country

    WHERE ak.n = 1
      AND bk.n = 1
)

SELECT * FROM positive_s2
UNION ALL
SELECT * FROM positive_s3
"""

silver_positive = con.sql(query).df()

print("Silver positive pairs:", len(silver_positive))
print("\nBy source:")
print(silver_positive["source"].value_counts())

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Silver positive pairs: 164009

By source:
source
S2    159188
S3      4821
Name: count, dtype: int64


In [5]:
silver_positive.head(20)

,source1_entity_id,candidate_entity_id,source
0,S1-886405942,S2-138046867,S2
1,S1-464464519,S2-378707035,S2
2,S1-709534371,S2-223038034,S2
3,S1-545433913,S2-839616035,S2
4,S1-435071560,S2-623817123,S2
5,S1-410561274,S2-696635067,S2
6,S1-249244306,S2-974399068,S2
7,S1-505811975,S2-175566180,S2
8,S1-925840440,S2-9962270,S2
9,S1-780559778,S2-533197229,S2


In [6]:
SILVER_DIR = "/content/Amazon-ML-Challenge/data/silver"

!mkdir -p "$SILVER_DIR"

silver_positive.to_parquet(
    f"{SILVER_DIR}/silver_positive.parquet",
    index=False
)

print("Saved silver positives.")

Saved silver positives.


In [ ]:
import duckdb
import pandas as pd

con = duckdb.connect()

# silver_positive already exists from the previous step
# Columns:
# source1_entity_id, candidate_entity_id, source

con.register("silver_positive_df", silver_positive)

DATA_DIR = "/content/Amazon-ML-Challenge/data/clean_data"

S1 = f"{DATA_DIR}/clean_source1.parquet"
S2 = f"{DATA_DIR}/clean_source2.parquet"
S3 = f"{DATA_DIR}/clean_source3.parquet"


# ============================================================
# FUNCTION TO GENERATE NEGATIVES FOR ONE TARGET SOURCE
# ============================================================

def generate_negatives(target_path, target_source, n_negatives=2):

    query = f"""
    WITH

    -- Source 1 records involved in our silver positives
    s1 AS (
        SELECT
            entity_id,
            country
        FROM read_parquet('{S1}')
    ),

    -- Target source
    target AS (
        SELECT
            entity_id,
            country
        FROM read_parquet('{target_path}')
        WHERE country IS NOT NULL
    ),

    -- Deterministic ordering inside each country
    target_ranked AS (
        SELECT
            entity_id,
            country,

            ROW_NUMBER() OVER (
                PARTITION BY country
                ORDER BY hash(entity_id)
            ) - 1 AS rn

        FROM target
    ),

    target_counts AS (
        SELECT
            country,
            COUNT(*) AS n
        FROM target
        GROUP BY country
    ),

    -- Only silver positives belonging to this source
    positives AS (
        SELECT
            source1_entity_id,
            candidate_entity_id AS positive_entity_id,
            source
        FROM silver_positive_df
        WHERE source = '{target_source}'
    ),

    base AS (
        SELECT
            p.source1_entity_id,
            p.positive_entity_id,
            p.source,

            s.country,
            tc.n AS target_count

        FROM positives p

        JOIN s1 s
          ON s.entity_id = p.source1_entity_id

        JOIN target_counts tc
          ON tc.country = s.country
    ),

    -- Generate several possible negative candidates.
    -- We use more attempts than needed so that candidates
    -- accidentally overlapping silver positives can be skipped.
    attempts AS (
        SELECT
            b.source1_entity_id,
            b.positive_entity_id,
            b.source,
            b.country,
            r.attempt

        FROM base b

        CROSS JOIN range(0, 10) AS r(attempt)
    ),

    candidate_negatives AS (
        SELECT
            a.source1_entity_id,
            a.positive_entity_id,
            a.source,
            t.entity_id AS candidate_entity_id,
            a.attempt

        FROM attempts a

        JOIN target_ranked t
          ON t.country = a.country
         AND t.rn = (
             hash(
                 concat(
                     a.source1_entity_id,
                     '|',
                     a.positive_entity_id,
                     '|',
                     '{target_source}',
                     '|',
                     CAST(a.attempt AS VARCHAR)
                 )
             ) % (
                 SELECT COUNT(*)
                 FROM target_counts tc
                 WHERE tc.country = a.country
             )
         )

        -- Candidate cannot be the known positive
        WHERE t.entity_id <> a.positive_entity_id
    ),

    -- Remove known silver-positive pairs
    filtered AS (
        SELECT DISTINCT
            c.source1_entity_id,
            c.candidate_entity_id,
            c.source,
            c.attempt

        FROM candidate_negatives c

        LEFT JOIN silver_positive_df p
          ON p.source1_entity_id = c.source1_entity_id
         AND p.candidate_entity_id = c.candidate_entity_id
         AND p.source = c.source

        WHERE p.source1_entity_id IS NULL
    ),

    numbered AS (
        SELECT
            source1_entity_id,
            candidate_entity_id,
            source,

            ROW_NUMBER() OVER (
                PARTITION BY source1_entity_id, source
                ORDER BY attempt
            ) AS neg_rank

        FROM filtered
    )

    SELECT
        source1_entity_id,
        candidate_entity_id,
        source,
        0 AS label

    FROM numbered

    WHERE neg_rank <= {n_negatives}
    """

    return con.sql(query).df()


# ============================================================
# GENERATE NEGATIVES
# ============================================================

print("Generating S2 negatives...")
neg_s2 = generate_negatives(
    S2,
    "S2",
    n_negatives=2
)

print("S2 negatives:", len(neg_s2))

print("\nGenerating S3 negatives...")
neg_s3 = generate_negatives(
    S3,
    "S3",
    n_negatives=2
)

print("S3 negatives:", len(neg_s3))


# ============================================================
# COMBINE
# ============================================================

negative_pairs = pd.concat(
    [neg_s2, neg_s3],
    ignore_index=True
)

print("\nTotal negative pairs:", len(negative_pairs))

print("\nNegatives by source:")
print(negative_pairs["source"].value_counts())

Generating S2 negatives...


In [ ]:
positive_pairs = silver_positive.copy()

positive_pairs["label"] = 1

positive_pairs = positive_pairs[
    [
        "source1_entity_id",
        "candidate_entity_id",
        "source",
        "label"
    ]
]

negative_pairs = negative_pairs[
    [
        "source1_entity_id",
        "candidate_entity_id",
        "source",
        "label"
    ]
]

train_pairs = pd.concat(
    [positive_pairs, negative_pairs],
    ignore_index=True
)

train_pairs = train_pairs.drop_duplicates(
    subset=[
        "source1_entity_id",
        "candidate_entity_id",
        "source"
    ]
)

print("Total training pairs:", len(train_pairs))
print("\nLabels:")
print(train_pairs["label"].value_counts())

print("\nSources:")
print(train_pairs["source"].value_counts())

In [ ]:
SILVER_DIR = "/content/drive/MyDrive/Amazon-ML-Challenge"

!mkdir -p "$SILVER_DIR"

train_pairs.to_parquet(
    f"{SILVER_DIR}/silver_training_pairs.parquet",
    index=False
)

print("Saved:")
print(f"{SILVER_DIR}/silver_training_pairs.parquet")

In [2]:
!git clone https://github.com/Parth-Kamal/Amazon-ML-Challenge.git

Cloning into 'Amazon-ML-Challenge'...
remote: Enumerating objects: 65, done.
remote: Counting objects: 100% (65/65), done.
remote: Compressing objects: 100% (50/50), done.
remote: Total 65 (delta 21), reused 39 (delta 11), pack-reused 0 (from 0)
Receiving objects: 100% (65/65), 58.94 KiB | 1.90 MiB/s, done.
Resolving deltas: 100% (21/21), done.


In [3]:
!ls -lah /content/Amazon-ML-Challenge

total 88K
drwxr-xr-x 6 root root 4.0K Sep 26 20:03 .
drwxr-xr-x 1 root root 4.0K Sep 26 20:03 ..
drwxr-xr-x 3 root root 4.0K Sep 26 20:03 code
-rw-r--r-- 1 root root  35K Sep 26 20:03 DATASET_INVENTORY.md
-rw-r--r-- 1 root root 2.2K Sep 26 20:03 Documentation_template.md
drwxr-xr-x 8 root root 4.0K Sep 26 20:03 .git
-rw-r--r-- 1 root root  237 Sep 26 20:03 .gitignore
-rw-r--r-- 1 root root 1.1K Sep 26 20:03 LICENSE
drwxr-xr-x 2 root root 4.0K Sep 26 20:03 notebooks
-rw-r--r-- 1 root root  15K Sep 26 20:03 README.md
-rw-r--r-- 1 root root    0 Sep 26 20:03 submissions.md
drwxr-xr-x 2 root root 4.0K Sep 26 20:03 utils


In [4]:
%cd /content/Amazon-ML-Challenge

/content/Amazon-ML-Challenge


In [5]:
!ls

code		      Documentation_template.md  notebooks  submissions.md
DATASET_INVENTORY.md  LICENSE			 README.md  utils


In [6]:
import os

print("Current directory:")
print(os.getcwd())

print("\nRepository:")
print(os.listdir("/content/Amazon-ML-Challenge"))

Current directory:
/content/Amazon-ML-Challenge

Repository:
['.gitignore', '.git', 'code', 'utils', 'notebooks', 'README.md', 'DATASET_INVENTORY.md', 'LICENSE', 'Documentation_template.md', 'submissions.md']


In [9]:
!find /content -maxdepth 4 -type d -name "dataset" 2>/dev/null

In [10]:
%cd /content/Amazon-ML-Challenge

/content/Amazon-ML-Challenge


In [12]:
!pwd

/content/Amazon-ML-Challenge


In [13]:
import os

print("Current working directory:")
print(os.getcwd())

print("\nRepository contents:")
print(os.listdir("/content/Amazon-ML-Challenge"))

print("\nSearching for dataset folders:")
!find /content -maxdepth 5 -type d -name "dataset" 2>/dev/null

Current working directory:
/content/Amazon-ML-Challenge

Repository contents:
['.gitignore', '.git', 'code', 'utils', 'notebooks', 'README.md', 'DATASET_INVENTORY.md', 'LICENSE', 'Documentation_template.md', 'submissions.md']

Searching for dataset folders:
